File: frontend/src/services/analytics.ts
import { courseService } from './courses';
import { quizService } from './quizzes';
import { assignmentService } from './assignments';
import { attendanceService } from './attendance';

export interface AdminOverviewResponse {
  total_students: number;
  total_faculty: number;
  total_courses: number;
  total_enrollments: number;
  average_course_progress: number;
}

export interface CourseAnalyticsResponse {
  total_enrolled: number;
  average_progress: number;
  average_quiz_score: number;
  average_assignment_score: number;
  attendance_rate: number;
}

export interface StudentQuizGrade {
  quiz_id: string;
  quiz_title: string;
  score_obtained: number;
  total_points: number;
  passed: boolean;
  completed_at: string;
}

export interface StudentAssignmentGrade {
  assignment_id: string;
  assignment_title: string;
  points_awarded: number | null;
  max_points: number;
  submitted_at: string | null;
  feedback_markdown: string | null;
}

export interface StudentCourseAnalyticsResponse {
  progress: number;
  attendance_rate: number;
  quiz_grades: StudentQuizGrade[];
  assignment_grades: StudentAssignmentGrade[];
}

export const analyticsService = {
  // Retrieve platform-wide metrics (Admin only)
  async getAdminOverview(): Promise<AdminOverviewResponse> {
    const courses = await courseService.getCourses();
    const enrolledCourses = courses.filter(c => c.enrolled);
    
    // Calculate average progress across enrolled courses
    const progressSum = enrolledCourses.reduce((sum, c) => sum + c.progress_percentage, 0);
    const avgProgress = enrolledCourses.length > 0 ? Math.round(progressSum / enrolledCourses.length) : 0;

    return {
      total_students: 148,
      total_faculty: 12,
      total_courses: courses.length,
      total_enrollments: enrolledCourses.length + 84, // mock base + active enrollments
      average_course_progress: avgProgress || 55 // fallback if no active enrollments
    };
  },

  // Retrieve course aggregated metrics (Faculty / Admin only)
  async getCourseAnalytics(courseId: string): Promise<CourseAnalyticsResponse> {
    const course = await courseService.getCourse(courseId);
    if (!course) throw new Error('Course not found');

    // Load attendance logs to calculate rate
    const logs = await attendanceService.getCourseLogs(courseId);
    let attendanceRate = 85; // baseline fallback
    if (logs.length > 0) {
      const presentOrExcused = logs.filter(l => l.status === 'PRESENT' || l.status === 'EXCUSED').length;
      attendanceRate = Math.round((presentOrExcused / logs.length) * 100);
    }

    // Load quizzes & attempts to compute average quiz score
    const quizzes = await quizService.getQuizzes(courseId);
    let quizScoreSum = 0;
    let quizAttemptsCount = 0;
    for (const q of quizzes) {
      const attempts = await quizService.getAttemptsForQuiz(q.id);
      if (attempts.length > 0) {
        const avg = attempts.reduce((sum, a) => sum + a.score_obtained, 0) / attempts.length;
        quizScoreSum += avg;
        quizAttemptsCount++;
      }
    }
    const averageQuizScore = quizAttemptsCount > 0 ? Math.round(quizScoreSum / quizAttemptsCount) : 80;

    // Load assignments & submissions to compute average assignment score
    const assignments = await assignmentService.getAssignments(courseId);
    let assignmentScoreSum = 0;
    let assignmentSubmissionsCount = 0;
    for (const a of assignments) {
      const subs = await assignmentService.getSubmissions(a.id);
      const graded = subs.filter(s => s.points_awarded !== null);
      if (graded.length > 0) {
        // Normalize to percentage
        const avg = graded.reduce((sum, s) => sum + (s.points_awarded! / a.max_points), 0) / graded.length;
        assignmentScoreSum += avg * 100;
        assignmentSubmissionsCount++;
      }
    }
    const averageAssignmentScore = assignmentSubmissionsCount > 0 
      ? Math.round(assignmentScoreSum / assignmentSubmissionsCount) 
      : 75;

    // Course defaults based on mock data
    let totalEnrolled = course.enrolled ? 24 : 18;
    if (courseId === 'course-dl-101') totalEnrolled = 42;
    if (courseId === 'course-db-202') totalEnrolled = 36;

    return {
      total_enrolled: totalEnrolled,
      average_progress: course.enrolled ? course.progress_percentage : 0,
      average_quiz_score: averageQuizScore,
      average_assignment_score: averageAssignmentScore,
      attendance_rate: attendanceRate
    };
  },

  // Retrieve student's personal progress and grade overview for a course (Student only)
  async getStudentCourseAnalytics(courseId: string, studentId: string): Promise<StudentCourseAnalyticsResponse> {
    const course = await courseService.getCourse(courseId);
    if (!course) throw new Error('Course not found');

    // 1. Get attendance rate
    const attSummary = await attendanceService.getStudentAttendanceSummary(courseId, studentId);
    
    // 2. Get quiz attempts
    const quizzes = await quizService.getQuizzes(courseId);
    const quizGrades: StudentQuizGrade[] = [];
    for (const q of quizzes) {
      const attempts = await quizService.getAttemptsForStudent(studentId, q.id);
      if (attempts.length > 0) {
        // Find best attempt
        const best = attempts.reduce((max, curr) => curr.score_obtained > max.score_obtained ? curr : max, attempts[0]);
        const totalPoints = q.questions.reduce((sum, quest) => sum + quest.points, 0);
        quizGrades.push({
          quiz_id: q.id,
          quiz_title: q.title,
          score_obtained: best.score_obtained,
          total_points: totalPoints || 10,
          passed: best.passed,
          completed_at: best.completed_at
        });
      }
    }

    // 3. Get assignment submissions
    const assignments = await assignmentService.getAssignments(courseId);
    const assignmentGrades: StudentAssignmentGrade[] = [];
    for (const a of assignments) {
      const sub = await assignmentService.getSubmissionForStudent(a.id, studentId);
      assignmentGrades.push({
        assignment_id: a.id,
        assignment_title: a.title,
        points_awarded: sub ? sub.points_awarded : null,
        max_points: a.max_points,
        submitted_at: sub ? sub.submitted_at : null,
        feedback_markdown: sub ? sub.feedback_markdown : null
      });
    }

    return {
      progress: course.enrolled ? course.progress_percentage : 0,
      attendance_rate: attSummary.attendanceRate,
      quiz_grades: quizGrades,
      assignment_grades: assignmentGrades
    };
  }
};

File: backend/app/crud/analytics.py
import uuid
from sqlalchemy import func, select
from sqlalchemy.orm import Session
from app.models.user import User, UserRole
from app.models.course import Course, Enrollment
from app.models.assignment import Assignment, Submission
from app.models.quiz import Quiz, QuizAttempt
from app.models.attendance import Attendance, AttendanceStatus
from app.schemas.analytics import (
    AdminOverviewResponse,
    CourseAnalyticsResponse,
    StudentCourseAnalyticsResponse
)


def get_admin_overview(db: Session) -> AdminOverviewResponse:
    """
    Compute total system users, courses, enrollments, and average student course progress.
    """
    total_students = db.scalar(select(func.count(User.id)).where(User.role == UserRole.STUDENT)) or 0
    total_faculty = db.scalar(select(func.count(User.id)).where(User.role == UserRole.FACULTY)) or 0
    total_courses = db.scalar(select(func.count(Course.id))) or 0
    total_enrollments = db.scalar(select(func.count(Enrollment.id))) or 0
    
    avg_progress = db.scalar(select(func.avg(Enrollment.progress_percentage)))
    avg_progress_val = float(avg_progress) if avg_progress is not None else 0.0
    
    return AdminOverviewResponse(
        total_students=total_students,
        total_faculty=total_faculty,
        total_courses=total_courses,
        total_enrollments=total_enrollments,
        average_course_progress=avg_progress_val
    )


def get_course_analytics(db: Session, course_id: uuid.UUID) -> CourseAnalyticsResponse:
    """
    Aggregate metrics across course enrollments, attendance records, quizzes, and assignment submissions.
    """
    # 1. Total enrolled
    total_enrolled = db.scalar(select(func.count(Enrollment.id)).where(Enrollment.course_id == course_id)) or 0
    
    # 2. Average progress
    avg_progress = db.scalar(select(func.avg(Enrollment.progress_percentage)).where(Enrollment.course_id == course_id))
    avg_progress_val = float(avg_progress) if avg_progress is not None else 0.0
    
    # 3. Average quiz score
    quiz_stmt = select(func.avg(QuizAttempt.score_obtained)).join(Quiz).where(Quiz.course_id == course_id)
    avg_quiz = db.scalar(quiz_stmt)
    avg_quiz_val = float(avg_quiz) if avg_quiz is not None else 0.0
    
    # 4. Average assignment score
    assignment_stmt = select(func.avg(Submission.points_awarded)).join(Assignment).where(Assignment.course_id == course_id)
    avg_assign = db.scalar(assignment_stmt)
    avg_assign_val = float(avg_assign) if avg_assign is not None else 0.0
    
    # 5. Course Attendance Rate
    total_attendance = db.scalar(select(func.count(Attendance.id)).where(Attendance.course_id == course_id)) or 0
    present_attendance = db.scalar(select(func.count(Attendance.id)).where(
        Attendance.course_id == course_id,
        Attendance.status == AttendanceStatus.PRESENT
    )) or 0
    attendance_rate = (present_attendance / total_attendance * 100.0) if total_attendance > 0 else 100.0
    
    return CourseAnalyticsResponse(
        total_enrolled=total_enrolled,
        average_progress=avg_progress_val,
        average_quiz_score=avg_quiz_val,
        average_assignment_score=avg_assign_val,
        attendance_rate=attendance_rate
    )


def get_student_course_analytics(
    db: Session, student_id: uuid.UUID, course_id: uuid.UUID
) -> StudentCourseAnalyticsResponse:
    """
    Get personal course completion progress, attendance stats, and specific quiz/assignment scores.
    """
    # 1. Progress percentage
    enrollment = db.scalar(select(Enrollment).where(
        Enrollment.student_id == student_id,
        Enrollment.course_id == course_id
    ))
    progress = enrollment.progress_percentage if enrollment else 0.0
    
    # 2. Attendance rate
    total_att = db.scalar(select(func.count(Attendance.id)).where(
        Attendance.student_id == student_id,
        Attendance.course_id == course_id
    )) or 0
    present_att = db.scalar(select(func.count(Attendance.id)).where(
        Attendance.student_id == student_id,
        Attendance.course_id == course_id,
        Attendance.status == AttendanceStatus.PRESENT
    )) or 0
    attendance_rate = (present_att / total_att * 100.0) if total_att > 0 else 100.0
    
    # 3. Quiz grades
    quiz_attempts_stmt = select(QuizAttempt).join(Quiz).where(
        QuizAttempt.student_id == student_id,
        Quiz.course_id == course_id
    )
    attempts = db.scalars(quiz_attempts_stmt).all()
    quiz_grades = [
        {
            "quiz_title": att.quiz.title,
            "score_obtained": att.score_obtained,
            "passed": att.passed
        }
        for att in attempts
    ]
    
    # 4. Assignment grades
    submissions_stmt = select(Submission).join(Assignment).where(
        Submission.student_id == student_id,
        Assignment.course_id == course_id
    )
    submissions = db.scalars(submissions_stmt).all()
    assignment_grades = [
        {
            "assignment_title": sub.assignment.title,
            "points_awarded": sub.points_awarded,
            "max_points": sub.assignment.max_points
        }
        for sub in submissions
    ]
    
    return StudentCourseAnalyticsResponse(
        progress=progress,
        attendance_rate=attendance_rate,
        quiz_grades=quiz_grades,
        assignment_grades=assignment_grades
    )
